toolkit/offchain/src/register.rs

```diff
@@ -0,0 +1,362 @@
+		let amount_builder = TransactionOutputBuilder::new()
+			.with_address(&validator.address(ctx.network))
+			.with_plutus_data(&datum)
+			.next()?;
+		let output = amount_builder.with_coin(&BigNum::zero()).build()?;
+		let min_ada = MinOutputAdaCalculator::new(
+			&output,
+			&DataCost::new_coins_per_byte(
+				&ctx.protocol_parameters.min_utxo_deposit_coefficient.into(),
+			),
+		)
+		.calculate_ada()?;
+		let output = amount_builder.with_coin(&min_ada).build()?;
+		tx_builder.add_output(&output)?;
+	}
+
+	tx_builder.balance_update_and_build(ctx)
+}
+
+#[cfg(test)]
+mod tests {
+	use super::register_tx;
+	use crate::csl::{OgmiosUtxoExt, TransactionContext};
+	use crate::test_values::{self, *};
+	use cardano_serialization_lib::{
+		Address, BigNum, ExUnits, NetworkIdKind, Transaction, TransactionInputs,
+	};
+	use ogmios_client::types::OgmiosValue;
+	use ogmios_client::types::{OgmiosTx, OgmiosUtxo};
+	use partner_chains_plutus_data::registered_candidates::block_producer_registration_to_plutus_data;
+	use proptest::{
+		array::uniform32,
+		collection::{hash_set, vec},
+		prelude::*,
+	};
+
+	use sidechain_domain::{
+		AdaBasedStaking, AuraPublicKey, BlockProducerRegistration, GrandpaPublicKey,
+		MainchainAddressHash, MainchainSignature, McTxHash, SidechainPublicKey, SidechainSignature,
+		UtxoId, UtxoIndex,
+	};
+
+	fn sum_lovelace(utxos: &[OgmiosUtxo]) -> u64 {
+		utxos.iter().map(|utxo| utxo.value.lovelace).sum()
+	}
+
+	const MIN_UTXO_LOVELACE: u64 = 1000000;
+	const FIVE_ADA: u64 = 5000000;
+
+	fn block_producer_registration(consumed_input: UtxoId) -> BlockProducerRegistration {
+		BlockProducerRegistration {
+			stake_ownership: AdaBasedStaking {
+				pub_key: test_values::mainchain_pub_key(),
+				signature: MainchainSignature(Vec::new()),
+			},
+			sidechain_pub_key: SidechainPublicKey(Vec::new()),
+			sidechain_signature: SidechainSignature(Vec::new()),
+			consumed_input,
+			own_pkh: MainchainAddressHash([0; 28]),
+			aura_pub_key: AuraPublicKey(Vec::new()),
+			grandpa_pub_key: GrandpaPublicKey(Vec::new()),
+		}
+	}
+
+	fn lesser_payment_utxo() -> OgmiosUtxo {
+		make_utxo(1u8, 0, 1200000, &payment_addr())
+	}
+
+	fn greater_payment_utxo() -> OgmiosUtxo {
+		make_utxo(4u8, 1, 1200001, &payment_addr())
+	}
+
+	fn input_utxo() -> OgmiosUtxo {
+		make_utxo(11u8, 0, 1000000, &payment_addr())
+	}
+
+	fn validator_addr() -> Address {
+		Address::from_bech32("addr_test1wpha4546lvfcau5jsrwpht9h6350m3au86fev6nwmuqz9gqer2ung")
+			.unwrap()
+	}
+
+	#[test]
+	fn register_tx_regression_test() {
+		let ex_units = ExUnits::new(&0u32.into(), &0u32.into());
+		let payment_key_utxos = vec![lesser_payment_utxo(), greater_payment_utxo(), input_utxo()];
+		let ctx = TransactionContext {
+			payment_key: payment_key(),
+			payment_key_utxos: payment_key_utxos.clone(),
+			network: NetworkIdKind::Testnet,
+			protocol_parameters: protocol_parameters(),
+		};
+		let own_registration_utxos = vec![payment_key_utxos.get(1).unwrap().clone()];
```

Comment:
This is confusing, because you have chosen `greater_payment_utxo` as `input_utxo`. Use `get(2)` to use the value produced by `input_utxo()`.
Perhaps will cascade change to assertions further down.

Reply:
`own_registration_utxos` are the old registrations that have to be invalidated, while `input_utxo` is just whatever utxo that you use to prove that you can create this transaction. I wanted to make sure there is no overlap.
Could you please confirm if this makes sense?

Reply:
`input_utxo` - lets rename it to `registration_utxo` to be consistent with pc-contracts-cli and make it clearer, at least for us.
So, this `input_utxo` = `registration_utxo` = a chosen UTXO, it it in the signed message, so cannot be "whatever".

Yes, `own_registration_utxos` are the old registrations. Invalidated = spent in the tx we just create here.

Reply:
Side note: I really hope that on-chain code prevent malicious actors from spending UTXOs not sent by them.

Reply:
But it is `registration_utxo` in the current UI and documentation. So, if `BlockProducerRegistration` is a new thing in this code base, I would really like to see some consistency. 

It is indeed arbitrarily chosen, but not before calling `register` but before getting signatures.

Reply:
I agree that it should be consistent, I just wanted to provide the other alternative that I've seen. I made a commit that makes the name consistent across the code base, it might deserve its own PR.